---
title: "Barcode calling summary"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    df_print: paged
    highlight: tango
    number_sections: no
    theme: flatly
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: yes
---

```{r knitr, include = FALSE}
knitr::opts_chunk$set(
    echo           = TRUE,
    error          = FALSE,
    fig.align      = "center",
    message        = FALSE,
    warning        = FALSE,
    fig.width      = 10,
    fig.height     = 8
)
```

```{r logs}
log <- file(snakemake@log[[1]], open = "wt")
sink(log)
sink(log, type = "message")
```

```{r libraries}
library(Seurat)
library(tidyverse)
library(DropletUtils)
source("workflow/scripts/R/functions.R")
```

```{r define-functions}
get_larry_molinfo <- function(molinfo) {
  
  larry_bc_pos <- which(molinfo$feature.type == "Custom")

  larry_molinfo <- molinfo$data %>% 
    data.frame() %>% 
    dplyr::filter(gene %in% larry_bc_pos) %>% 
    mutate(larry_bc = molinfo$genes[gene]) %>% # Create col with larry bc name instead of idx from molinfo$genes
    mutate(larry_color = str_replace(larry_bc, "_.*$", "")) %>% # Add color information to split barcodes
    dplyr::select(-gene)
  
    return(larry_molinfo)
  
}

efficiency_and_integrations_by_threshold <- function(seurat, slot = "data", larry_assay = "Larry", thresholds = seq(0, 10, 0.1)) {
  require(tidyverse)

  m <- as(GetAssayData(object = seurat, slot = slot, assay = larry_assay), "dgTMatrix")
  n_cells <- length(Cells(seurat))

  df_bcs <- data.frame(
    larry    = m@Dimnames[[1]][m@i + 1], # Add +1 because the indexes of the dgTMatrix are 0-based
    cells    = m@Dimnames[[2]][m@j + 1], # Add +1 because the indexes of the dgTMatrix are 0-based
    norm_umi = m@x
  ) %>% mutate(color = str_replace(larry, "-.*$", ""))

  df <- map_df(
    thresholds,
    \(x) df_bcs %>%
      filter(norm_umi >= x) %>%
      group_by(color) %>%
      count(cells) %>%
      summarise(
        threshold = x,
        cells_bc = n() / n_cells,
        cells_multiple_bc = sum(n > 1) / n_cells
      )
  )

  df <- pivot_longer(df, -c(threshold, color))
  return(df)
}
```

```{r prepare-data, results = 'hide'}
seurat_rds    <- snakemake@input[[1]]
molec_info_h5 <- snakemake@params[["molecule_info"]]
  
seurat      <- readRDS(seurat_rds)
molinfo     <- read10xMolInfo(molec_info_h5,
                              get.cell    = TRUE,
                              get.umi     = TRUE,
                              get.gem     = FALSE,
                              get.gene    = TRUE,
                              get.reads   = TRUE,
                              get.library = FALSE
                              )

# Modify cell names form molinfo file to match those from seurat (basically add whatever is)
# before and after the cellular barcode (which consists in random 16 nucleotides, by now)
cell_prefix <- Cells(seurat) %>% str_replace("[AGTC]{16}-.*", "")
cell_suffix <- Cells(seurat) %>% str_replace(".*_[AGTC]{16}", "")

molinfo$data <- molinfo$data %>% 
  data.frame() %>%
  dplyr::mutate(cell = paste0(cell_prefix, cell, cell_suffix)) %>% 
  filter(cell %in% Cells(seurat)) %>% 
  DataFrame()

molinfo_larry <- get_larry_molinfo(molinfo)
```


# Histogram of UMI reads

Number of UMIs that have X amount of reads associated to them.

```{r, fig.width= 7.5, fig.height=5}
molinfo_larry %>% 
  ggplot(aes(reads)) + 
  geom_histogram(bins = 100) +
  scale_y_log10() +
  scale_x_log10() +
  theme_bw() +
  facet_wrap(~larry_color)
```

# Percentage of cells barcoded and multiple integrations

In the following plot you can see the proportion of cells barcoded by sample, and also the proportion of cells barcoded by sample in which +1 different barcode has been detected. The idea is to find the sweet spot between the maximum amount of barcoded cells possible with the minimum amount of multiple integrations.

## By sample

```{r clone-calling-threshold-sample, fig.width=7.5, fig.height=7}
for (i in unique(seurat$sample)) {
  
  tmp <- subset(seurat, subset = sample == i)
  
  print(glue("Plotting sample {i}:"))
  
  p <- efficiency_and_integrations_by_threshold(tmp, slot = "counts", larry_assay = "Larry_filt", thresholds = seq(1, 10, 1)) %>%
    ggplot(aes(x = threshold, y = value, color = name)) +
    geom_line() +
    geom_point() +
    facet_wrap(~color, scales = "free_y") +
    theme.2 +
    labs(x = "UMI threshold", y = "Proportion of cells") +
    scale_color_brewer(palette = "Dark2") +
    # geom_vline(xintercept = 7.5, linetype = "dashed") +
    theme(legend.position = "bottom")
  
  print(p)

}
```

## By subsample

If cellhashing was not perform when preparing the libraries, ignore this.

```{r clone-calling-threshold-subsample, fig.width=7.5, fig.height=7}
for (i in unique(seurat$subsample)) {
  
  tmp <- subset(seurat, subset = subsample == i)
  
  print(glue("Plotting sample {i}:"))
  
  p <- efficiency_and_integrations_by_threshold(tmp, slot = "counts", larry_assay = "Larry_filt", thresholds = seq(1, 10, 1)) %>%
    ggplot(aes(x = threshold, y = value, color = name)) +
    geom_line() +
    geom_point() +
    facet_wrap(~color, scales = "free_y") +
    theme.2 +
    labs(x = "UMI threshold", y = "Proportion of cells") +
    scale_color_brewer(palette = "Dark2") +
    # geom_vline(xintercept = 7.5, linetype = "dashed") +
    theme(legend.position = "bottom")
  
  print(p)

}
```